src/search3_indexer.erl

```diff
@@ -0,0 +1,221 @@
+                    doc_acc := [],
+                    search_seq => LastSeq,
+                    % make sure this reset here is correct
+                    last_seq := 0
+                }
+        end
+    end),
+    case State3 of
+        finished ->
+            % should we ret_os_process(Proc) here? or in the after clause
+            % in update/2?
+            ok;
+        _ ->
+            update_int(Db, State3)
+    end.
+
+fold_changes(State) ->
+    #{
+        search_seq := SearchSeq,
+        limit := Limit,
+        tx_db := TxDb
+    } = State,
+    fabric2_db:fold_changes(TxDb, SearchSeq,
+        fun load_changes/2, State, [{limit, Limit}]).
+
+% grabs the document from changes feed, and stores it into the document
+% accumulator for indexing later
+load_changes(Change, Acc) ->
+    #{
+        doc_acc := DocAcc,
+        count := Count,
+        tx_db := TxDb
+    } = Acc,
+    #{
+        id := Id,
+        sequence := LastSeq,
+        deleted := Deleted
+    } = Change,
+    Acc1 = case Id of
+        <<"_design/", _/binary>> ->
+            maps:merge(Acc, #{
+                count => Count + 1,
+                last_seq => LastSeq
+                });
+        _ ->
+            Doc = if Deleted -> []; true ->
+                case fabric2_db:open_doc(TxDb, Id) of
+                    {ok, Doc0} -> Doc0;
+                    {not_found, _} -> []
+                end
+            end,
+            Change1 = maps:put(doc, Doc, Change),
+            maps:merge(Acc, #{
+                doc_acc => DocAcc ++ [Change1],
+                count => Count + 1,
+                last_seq => LastSeq
+            })
+    end,
+    {ok, Acc1}.
+
+index_docs(Index, Proc, Seq, PurgeSeq, Docs) ->
+    DocIndexerFun = fun
+        (#{deleted := true, id:= Id}) ->
+            search3_rpc:delete_index(Index, Id, Seq, PurgeSeq);
+        (Change) ->
+            #{doc := Doc, id:= Id} = Change,
+            Fields = extract_fields(Proc, Doc),
+            search3_rpc:update_index(Index, Id, Seq, PurgeSeq, Fields)
+    end,
+    lists:foreach(DocIndexerFun, Docs),
+    ok.
+
+extract_fields(Proc, Doc) ->
+    Json = couch_doc:to_json_obj(Doc, []),
+    [Fields|_] = proc_prompt(Proc, [<<"index_doc">>, Json]),
+    [list_to_tuple(Field) || Field <- Fields].
+
+report_progress(State, UpdateType) ->
+    #{
+        tx_db := TxDb,
+        job := Job,
+        job_data := JobData,
+        last_seq := LastSeq
+    } = State,
+
+    #{
+        <<"db_name">> := DbName,
+        <<"ddoc_id">> := DDocId,
+        <<"name">> := IndexName,
+        <<"sig">> := Sig
+    } = JobData,
+
+    % Reconstruct from scratch to remove any
+    % possible existing error state.
+    NewData = #{
+        <<"db_name">> => DbName,
+        <<"ddoc_id">> => DDocId,
+        <<"name">> => IndexName,
+        <<"sig">> => Sig,
+        <<"search_seq">> => LastSeq
+    },
+
+    case UpdateType of
+        update ->
+            couch_jobs:update(TxDb, Job, NewData);
```

Comment:
Here make sure to check return value of update. The return will be `{ok, Job}` or `{error, ...}`. For `{ok, Job}` you'd want to perhaps update the Job object in the state with this new value. (couch_views forgot to do it, it will have to be updated as well).

Then we'd definitely want to check for `{error, halt}` if we get that back we'd know the job had timed out and was re-enqueued, and possibly running on another pod. In that case we'd want to log an error and exit. At some point if we get too many of those errors in the log we'd know to either change the update strategy (how often it happens) or increase the timeout.